modules/ximgproc/src/weighted_median_filter.cpp

```diff
@@ -0,0 +1,723 @@
+    sort(data,data+alls);
+
+    // Find lower bound and upper bound of the pixel values
+    double maxVal,minVal;
+    minMaxLoc(img,&minVal,&maxVal);
+    float maxRange = (float)(maxVal - minVal);
+    float th = 1e-5f;
+
+    float l = 0, r = maxRange*2.0f/nI;
+    // Perform binary search on error bound
+    while(r-l > th)
+    {
+        float m = (r+l)*0.5f;
+        bool suc = true;
+        float base = (float)minVal;
+        int cnt=0;
+        for(int i=0;i<alls;i++)
+        {
+            if(data[i].first>base+m)
+            {
+                cnt++;
+                base = data[i].first;
+                if(cnt==nI)
+                {
+                    suc = false;
+                    break;
+                }
+            }
+        }
+        if(suc)r=m;
+        else l=m;
+    }
+
+    Mat retImg(img.size(),CV_32SC1);
+    int *retImgPtr = retImg.ptr<int>();
+
+    // In the sorted list, divide pixel values into clusters according to the minimum error bound
+    // Quantize each value to the median of its cluster
+    // Also record the mapping of quantized value and quantized index.
+    float base = (float)minVal;
+    int baseI = 0;
+    int cnt = 0;
+    for(int i=0;i<=alls;i++)
+    {
+        if(i==alls || data[i].first>base+r)
+        {
+            mapping[cnt] = data[(baseI+i-1)>>1].first; //median
+            if(i==alls)break;
+            cnt++;
+            base = data[i].first;
+            baseI = i;
+        }
+        retImgPtr[data[i].second] = cnt;
+    }
+
+    free(data);
+
+    //end of the function
+    outImg = retImg;
+}
+
+/***************************************************************/
+/* Function: from32STo32F
+ * Description: convert the quantization index image back to the floating-point image accroding to the mapping
+***************************************************************/
+void from32STo32F(Mat &img, Mat &outImg, float *mapping)
+{
+    Mat retImg(img.size(),CV_32F);
+    int rows = img.rows, cols = img.cols, alls = rows*cols;
+    float *retImgPtr = retImg.ptr<float>();
+    int *imgPtr = img.ptr<int>();
+
+    // convert 32S index to 32F real value
+    for(int i=0;i<alls;i++)
+    {
+        retImgPtr[i] = mapping[imgPtr[i]];
+    }
+
+    // end of the function
+    outImg = retImg;
+}
+
+
+/***************************************************************
+ * Function: float2D
+ * Description: allocate a 2D float array with dimension "dim1 x dim2"
+ ***************************************************************/
+float** float2D(int dim1, int dim2)
+{
+    float **ret = new float*[dim1];
+    ret[0] = new float[dim1*dim2];
+    for(int i=1;i<dim1;i++)ret[i] = ret[i-1]+dim2;
+
+    return ret;
+}
+
+/***************************************************************
+ * Function: float2D_release
+ * Description: deallocate the 2D array created by float2D()
+ ***************************************************************/
+void float2D_release(float **p)
+{
+    delete []p[0];
```

Comment:
There is a [compiler warning](http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_android/builds/1581/steps/compile%20release/logs/warnings%20%281%29) ([full log](http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_android/builds/1581/steps/compile%20release/logs/stdio))